presto-spark-base/src/main/java/com/facebook/presto/spark/execution/common/Properties.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spark.execution.common;
+
+import com.google.common.collect.ImmutableMap;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * A class that stores key value properties. It also provides {@link PropertiesPopulator} interfaces of populating/materializing the properties. See {@link FilePropertiesPopulator}
+ * for a concrete implementation of a {@link PropertiesPopulator}.
+ */
+public class Properties
+{
+    private PropertiesPopulator populator;
+    private ImmutableMap<String, String> propertiesMap;
+
+    public Properties(PropertiesPopulator populator, ImmutableMap<String, String> propertiesMap)
+    {
+        this.populator = populator;
+        this.propertiesMap = propertiesMap;
+    }
+
+    public Map<String, String> getProperties()
+    {
+        return propertiesMap;
+    }
+
+    public void populateProperties()
+    {
+        populator.populate(propertiesMap);
+    }
+
+    public abstract static class PropertiesBuilder<T extends PropertiesBuilder<T>>
+    {
+        protected final HashMap<String, String> propertiesMap = new HashMap<>();
+        private PropertiesPopulator populator;
+
+        public T setPopulator(PropertiesPopulator populator)
+        {
+            this.populator = populator;
+            return getThis();
+        }
+
+        public T setProperty(String propertyName, String value)
+        {
+            propertiesMap.put(propertyName, value);
+            return getThis();
+        }
+
+        public abstract T getThis();
```

Comment:
could you explain more about why we make `getThis()` `abstract` rather than just return this? From the below implementation, I don't see any place we have any different implementation.

Reply:
Yeah this is a bit confusing. The whole point is abstraction on builder class so that common builder methods can be implemented only once in this class. 

Note the builder class is of generic type PropertiesBuilder<T extends PropertiesBuilder<T>>, this way through any builder methods we can directly get back the child type (T) builder so that we can call its' unique methods depending on what the actual builder implementations are.  

So above is the background. Now why using getThis() instead of "this" directly:
"this" keyword from this generic class is going to return PropertiesBuilder<T> which is the parent type not what's intended (child type T). So we need an interface getThis() for each child class to implement to return the actual child type.

I'm not sure if I have given clarifications through above words since it is indeed a bit convoluted also for me to give a clear explanation.. Let me know if further chats are needed.

Reply:
To return the child type T instead of parent, you can do this:
```
        public T setPopulator(PropertiesPopulator populator)
        {
            this.populator = populator;
            return (T) this;
        }
```

Reply:
To return the child type T instead of parent, you can do this:

```
        public T setPopulator(PropertiesPopulator populator)
        {
            this.populator = populator;
            return (T) this;
        }
```